gnc/sub8_perception/nodes/bins_2d.py

```diff
@@ -0,0 +1,173 @@
+        if self.last_draw_image is not None:
+            self.image_pub.publish(self.last_draw_image)
+
+    def image_cb(self, image):
+        '''Hang on to last image'''
+        self.last_image = image
+        self.last_image_time = self.image_sub.last_image_time
+        if self.camera_model is None:
+            if self.image_sub.camera_info is None:
+                return
+
+            self.camera_model = image_geometry.PinholeCameraModel()
+            self.camera_model.fromCameraInfo(self.image_sub.camera_info)
+
+    def find_single_bin(self, img, bin_type):
+        """Find the bins and their orientations."""
+        assert bin_type in self.bins[bin_type], "Bins_2d does not know bin color: {}".format(bin_type)
+        if img is not None:
+
+            debug_image = np.copy(img)
+            img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+            zeros = np.array([0])
+            mask = cv2.inRange(img, 127, 255)
+            img = cv2.bitwise_or(mask, zeros)
+            contours, hierarchy = cv2.findContours(img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+            contours = contour_sort(contours)
+
+            """This finds the bins and looks for the one that is orange or is not
+               orange. Each bin is given an orangeness rating and either the most 
+               or least orange bin is selected.
+            """
+            if len(contours) > 0:
+                bins = 4
+                if (self.bin_type == 'orange'):
+                    orangeness = 0
+                else:
+                    orangeness = 100000
+                if len(contours) < 4:
+                    bins = len(contours)
+                for i in range(0, bins):
+                    x, y, w, h = cv2.boundingRect(contours[i])
+                    roi = debug_image[y: y + h, x: x + w]
+                    temp = evaluate_bin(roi)
+                    if ((orangeness > temp and self.bin_type == 'norange') 
+                    or (orangeness < temp and self.bin_type == 'orange')):
+                        orangeness = temp
+                        M = cv2.moments(contours[i])
+                        cx = int(M['m10'] / M['m00'])
+                        cy = int(M['m01'] / M['m00'])
+                        img_h, img_w, _ = np.shape(debug_image)
+                        point = (cx, cy)
+                        (_, _), (_, _), rad = cv2.fitEllipse(contours[i])
+                    cv2.rectangle(debug_image, (x, y), (x + w, y + h), (127), 2)
+                    ellipse = cv2.fitEllipse(contours[i])
+                    cv2.ellipse(debug_image, ellipse, (170), 2)
+                cv2.circle(debug_image, point, 5, (0, 0, 255), -1)
+                point = [cx - (img_w / 2), cy - (img_h / 2)]
+                tuple_center = (point[0], point[1], 0)
+            self.last_draw_image = debug_image
+            rad = ((rad - 180) * np.pi) / 180.0
+            return tuple_center, rad
```

Comment:
Should the 3 lines here be tabbed in? If there are no contours find these variables are undefined I think